src/Compilers/CSharp/Portable/Compilation/SyntaxAndDeclarationManager.cs

```diff
@@ -565,6 +593,45 @@ public SyntaxAndDeclarationManager ReplaceSyntaxTree(SyntaxTree oldTree, SyntaxT
                 this.MessageProvider,
                 this.IsSubmission,
                 state);
+
+            OneOrMany<ImmutableSegmentedHashSet<string>> tryGetLastComputedMemberNames()
+            {
+                var previousRootNamespaceDeclaration = declMapBuilder[oldTree];
+                if (previousRootNamespaceDeclaration.IsValueCreated)
+                {
+                    // we computed the last root.  It will have up to date member names.  So just return those if present.
+                    var stack = s_declarationStack.Allocate();
+                    stack.Push(previousRootNamespaceDeclaration.Value);
+
+                    var builder = ArrayBuilder<ImmutableSegmentedHashSet<string>>.GetInstance();
+
+                    while (stack.Count > 0)
+                    {
+                        var current = stack.Pop();
+
+                        // As we walk down, push children on in reverse order so we get a DFS walk.
+
+                        for (int i = current.Children.Length - 1; i >= 0; i--)
+                            stack.Push(current.Children[i]);
+
+                        // Process any type we see before we process its nested types.
+                        if (current is SingleTypeDeclaration singleType && DeclarationTreeBuilder.CachesComputedMemberNames(singleType))
+                            builder.Add(singleType.MemberNames);
+                    }
+
+                    s_declarationStack.Free(stack);
+
+                    return builder.ToOneOrManyAndFree();
+                }
+
+                // The previous root wasn't computed yet.  So just return whatever info we have from before.
+                if (lastComputedMemberNamesMap.TryGetValue(oldTree, out var lastComputedMemberNames))
```

Comment:
If  results could be different, then how do we deal with the difference?

Reply:
> Why not check this first?

lastComputedMemberNamesMap might contain items for a version of the file several edits in the past.  We want to use the most recent data if possible.  If we can't (Because the flurry of edits came in too quick, and our data is more stale) we accept an older (but potentially more innacurate) version.

> If results could be different, then how do we deal with the difference?

All this work is simply an optimization to allow memory reuse in a common case. In any event where we are unable to do that, we simply pay the same price in memory that we do today without this change.

In other words: 

1. we're always going to try to compute the new member names for a type decl.
2. those member names are commonly going to be the same as the last state of the type decl.
3. in that case, we can literally just reuse the same member-names-collection, instead of allocating a new one that contains all the same names.

Reply:
Updated docs to explain things better (i hope) :)